### PR TITLE
Adds a function to grant research points for slime extracts

### PR DESCRIPTION
**Pull request by TheVekter**
## About The Pull Request
Another part of the R&D rework - Xenobiologists can now spend a slime extract by feeding it to the slime camera console to get research points for the _first instance of that extract_. This consumes the core in the process. As an example: if you use a green extract, you can't use another one for points. This maxes out at a pretty big chunk of points, but most rounds don't see every breed of slimes bred, so I don't expect it to be a major issue.

The points are as follows for each slime:

Tier 0 = 100 (Grey)
Tier 1 = 500 (Orange, purple, blue, metal)
Tier 2 = 1000 (Yellow, dark purple, dark blue, silver)
Tier 3 = 1500 (Bluespace, sepia, cerulean, pyrite)
Tier 4 = 2000 (Red, green, pink, gold)
Tier 5 = 2500 (Oil, black, light pink, adamantine
Rainbow = 5000

## Why It's Good For The Game

Xenobiology has ways to contribute to the station as a whole but no way to actively contribute to research. This fixes that. This should also help make up for a lack of the Experimentor while I work on the Research Scanner.

## Changelog
:cl:
add: Added the ability for Xenobiologists to spend slime extracts for research points. Only one of each extract can be used for points.
/:cl:

### CODE REVIEW

**Comment by TheVekter** [comment]
For anyone wondering, 22 types of slime = 22k research points throughout the round if _everything_ is bred. This will probably come with a passive gen nerf in the near future.

**Comment by knacker48** [comment]
So they all have the same point bonus? I was kind of expecting higher tiers of slimes to give more than lower tiers, or atleast for rainbow slimes to give more based on how hard they are to get.

That said I really like this change, pushes xenobio more into actively helping the station which is always good.

**Comment by TheVekter** [comment]
> 
> 
> So they all have the same point bonus? I was kind of expecting higher tiers of slimes to give more than lower tiers, or atleast for rainbow slimes to give more based on how hard they are to get.
> 
> That said I really like this change, pushes xenobio more into actively helping the station which is always good.

I'm having issues getting this to work. I wanted the rainbows to at least give more but it's being a pain about recognizing them. This is something I want to do, though, so I'll try to make it happen.

**Comment by XDTM** [comment]
You could assign a research value to all extracts. Maybe it could also be manipulated to make the less useful colors more worthwhile.

**Comment by TheVekter** [comment]
> 
> 
> You could assign a research value to all extracts. Maybe it could also be manipulated to make the less useful colors more worthwhile.

Not a bad idea. Shouldn't be too hard to do.

**Comment by trollbreeder** [comment]
What about crossbreeds? Do those count as separate extracts? (Example, you put in a light pink extract into the console, and then a stabilized light pink extract. Does the stabilized light pink extract give points in this situation?)

**Comment by XDTM** [comment]
Crossbreeds are coded as their own thing, so they'd not be researchable unless they're explicitly coded to do so

**Comment by TheVekter** [comment]
> 
> 
> What about crossbreeds? Do those count as separate extracts? (Example, you put in a light pink extract into the console, and then a stabilized light pink extract. Does the stabilized light pink extract give points in this situation?)

No plans to give crossbreeds a value - that's a massive waste for how powerful they are.

**Comment by trollbreeder** [comment]
> > What about crossbreeds? Do those count as separate extracts? (Example, you put in a light pink extract into the console, and then a stabilized light pink extract. Does the stabilized light pink extract give points in this situation?)
> 
> No plans to give crossbreeds a value - that's a massive waste for how powerful they are.

not a wasteful if you made really powerful crossbreeds the xenobio version of maxcaps
emergent gameplay

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 1]
```diff
@@ -169,6 +170,19 @@
 		current_potion = O
 		to_chat(user, "<span class='notice'>You load [O] in the console's potion slot[replaced ? ", replacing the one that was there before" : ""].</span>")
 		return
+	else if(istype(O, /obj/item/slime_extract))
+		var/obj/item/slime_extract/E = O
+		if(!already_researched[E.type])
+			playsound(src, 'sound/machines/ping.ogg', 50, 3, -1)
+			visible_message("<span class='notice'>You insert the [E] into a slot on the [src]. It pings and prints out some research notes worth [E.research] points!</span>")
```
Comment (Qustinnus) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Remove uses of "the" infront of vars. We handle those automatically when the name is used. Else it comes out as The the slime extract into the the console

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 2]
```diff
@@ -35,6 +35,7 @@
 	var/obj/item/slimepotion/slime/current_potion
 	var/max_slimes = 5
 	var/monkeys = 0
+	var/static/list/already_researched = list() //List of cores that've already been used for research
```
Comment (Qustinnus) [inline code/modules/research/xenobiology/xenobio_camera.dm]: I'm not a fan of this being saved in a static list on the xenobio camera, can we not use the research subsystem for this?

File: code/modules/research/xenobiology/xenobiology.dm [hunk 3]
```diff
@@ -1,3 +1,13 @@
+// Defines for research points //
+
+#define SLIME_TIER_0 100
+#define SLIME_TIER_1 500
+#define SLIME_TIER_2 1000
+#define SLIME_TIER_3 1500
+#define SLIME_TIER_4 2000
+#define SLIME_TIER_5 2500
+#define SLIME_TIER_RAINBOW 5000
+
```
Comment (Qustinnus) [inline code/modules/research/xenobiology/xenobiology.dm]: Move these to the _Defines folder and make them something more descriptive regarding their purpose.

File: code/modules/research/xenobiology/xenobio_camera.dm (hunk 2 above)
Comment (TheVekter) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Possibly, but this is the only place it's ever going to be referenced. It makes more sense to just tie it to the cameras IMO.

Comment (XDTM) [reply code/modules/research/xenobiology/xenobio_camera.dm]: The name is a bit generic for a static list, so if you're keeping it there i'd make it more specific.
I do second the opinion that it'd belong to the research subsystem though.

**Comment by TheVekter** [comment]
> 
> 
> > > What about crossbreeds? Do those count as separate extracts? (Example, you put in a light pink extract into the console, and then a stabilized light pink extract. Does the stabilized light pink extract give points in this situation?)
> > 
> > 
> > No plans to give crossbreeds a value - that's a massive waste for how powerful they are.
> 
> not a wasteful if you made really powerful crossbreeds the xenobio version of maxcaps
> emergent gameplay, use it to kill everyone on the station, to powergame, or to give the RnD team a high five after plopping in the charged rainbow

This might be changing in the future, but you guys are going to be pretty upset when the maxcap thing is gone in a few weeks.

**Comment by TheVekter** [comment]
Going a different direction with this, I think.

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 4]
```diff
@@ -169,6 +169,19 @@
...
+		var/obj/item/slime_extract/E = O
+		if(!SSresearch.slime_already_researched[E.type])
+			playsound(src, 'sound/machines/ping.ogg', 50, 3, -1)
+			visible_message("<span class='notice'>You insert [E] into a slot on the [src]. It pings and prints out some research notes worth [E.research] points!</span>")
+			new /obj/item/research_notes(drop_location(src), E.research, "xenobiology")
+			SSresearch.slime_already_researched[E.type] = TRUE
+			qdel(O)
+			return
+		else
+			visible_message("<span class='notice'>[src] buzzes and displays a message: Slime extract already researched!")
```
Comment (AnturK) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Needs `</span>`

File: code/modules/research/xenobiology/xenobiology.dm [hunk 5]
```diff
@@ -16,6 +16,7 @@
 	var/effectmod
 	var/list/activate_reagents = list() //Reagents required for activation
 	var/recurring = FALSE
+	var/research //Research point value
```
Comment (AnturK) [inline code/modules/research/xenobiology/xenobiology.dm]: If it defaults to none you should probably check if it has any value when trying to redeem it.

**Review by AnturK — CHANGES_REQUESTED** [review]
Looks alright. Should this be working with used up extracts ?

**Comment by TheVekter** [comment]
> Looks alright. Should this be working with used up extracts ?



> 
> 
> Looks alright. Should this be working with used up extracts ?

I was thinking probably not. I went ahead and coded it, just had to find something to hook it to so it could tell when they've been used up.

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 6]
```diff
@@ -169,6 +169,28 @@
 		current_potion = O
 		to_chat(user, "<span class='notice'>You load [O] in the console's potion slot[replaced ? ", replacing the one that was there before" : ""].</span>")
 		return
+	else if(istype(O, /obj/item/slime_extract))
+		var/obj/item/slime_extract/E = O
+		if(!SSresearch.slime_already_researched[E.type])
+			if(E.research == null || E.research == 0)
```
Comment (AnturK) [inline code/modules/research/xenobiology/xenobio_camera.dm]: You can just do `!E.research` here

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 7]
```diff
@@ -169,6 +169,28 @@
...
+				playsound(src, 'sound/machines/buzz-sigh.ogg', 50, 3, -1)
+				visible_message("<span class='notice'>[src] buzzes and displays a message: Invalid extract! (You shouldn't be seeing this. If you are, tell someone.)</span>")
+				return
+			if(E.Uses == 0)
```
Comment (AnturK) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Here <= 0 is probably better.

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 8]
```diff
@@ -169,6 +169,28 @@
...
+				playsound(src, 'sound/machines/buzz-sigh.ogg', 50, 3, -1)
+				visible_message("<span class='notice'>[src] buzzes and displays a message: Extract consumed - no research available.</span>")
+				return
+			else
+				playsound(src, 'sound/machines/ping.ogg', 50, 3, -1)
+				visible_message("<span class='notice'>You insert [E] into a slot on the [src]. It pings and prints out some research notes worth [E.research] points!</span>")
+				new /obj/item/research_notes(drop_location(src), E.research, "xenobiology")
```
Comment (AnturK) [inline code/modules/research/xenobiology/xenobio_camera.dm]: drop_location doesn't have any parameters, the src is unnecessary.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Pull request merged**